### DISCUSSION TITLE
How to provide types for a conditionally included relation?

### DISCUSSION THREAD

**Discussion by Xevion** [prisma/prisma#16514] [Q&A] [answered]
Here is the schema for my organization's Member table (simplified).

```prisma
model Member {
    id                 String      @id
    email              String      @unique
    data               MemberData?
}
```

Notice the `data` relation to the `MemberData` table. 

The following function exists in our codebase for fetching a specific user...

```ts
/**
 * Acquire a member given their unique identifier.
 * @param id
 * @param extendedData If true, include the extended member data.
 */
export async function getMember(id: string, extendedData: boolean = false) {
	return await prisma.member.findUnique({
		where: {
			id: id,
		},
		include: extendedData
			? {
					data: true,
			  }
			: undefined,
	});
}
```

The only issue with this function is that it does not provide type infereces for the `data` attribute, which is lacking from the generated `Member` type. See below, copy and pasted from the generated file containing it (simplified to match first provided schema).

```ts
export type Member = {
  id: string
  email: string
}
```

I've searched and scanned the 7,000 line file containing all of Prisma's generated types, but none of them have what I want. I've tried looking at the method declaration, but nothing there shows what I want. I'm pretty unsure of what the intended way of doing this is. The only solution I can think of is simply tacking it on with a union operator:

```ts
export async function getMember(
	id: string,
	extendedData: boolean = false
): Promise<Member | (Member & { data: MemberData }) | null> {
    ...
}
```

I thought this would work, but it's preventing access to the `data` property anyways. I found some workarounds, but I'm still wondering: What should I do here?

**Comment by nurul3101** [comment, marked as answer]
Hey @Xevion 👋 

From the schema file it looks like data is a relation field. In the generated types only the scalar fields are included and relation fields are not included by default, which explains why you are only seeing `id` and `email` in Member generated type.

If you wish to include the relations field then you can use Prisma.Validator API as described in this [section](https://www.prisma.io/docs/concepts/components/prisma-client/advanced-type-safety/operating-against-partial-structures-of-model-types#solution), can you have a look at it?

**Reply by Xevion** [reply]
Ah, I didn't know about this page or the Prisma.Validator API at all. Thanks for linking me that. A quick look shows that it's likely what I was after, so I'm considering this the 'answer' to my problem. 👍

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
